crates/factors/src/factor.rs

```diff
@@ -0,0 +1,163 @@
+use std::any::Any;
+
+use anyhow::Context;
+
+use crate::{
+    App, InstancePreparers, Linker, ModuleLinker, PrepareContext, RuntimeConfig, RuntimeFactors,
+};
+
+pub trait Factor: Any + Sized {
+    /// Per-app state for this factor.
+    ///
+    /// See [`Factor::configure_app`].
+    type AppState: Default;
+
+    /// The [`FactorInstancePreparer`] for this factor.
+    type InstancePreparer: Default;
+
+    /// The per-instance state for this factor, constructed by a
+    /// [`FactorInstancePreparer`] and available to any host-provided imports
+    /// defined by this factor.
+    type InstanceState;
+
+    /// Initializes this Factor for a runtime. This will be called at most once,
+    /// before any call to [`FactorInstancePreparer::new`]
+    fn init<T: RuntimeFactors>(&mut self, mut ctx: InitContext<T, Self>) -> anyhow::Result<()> {
+        // TODO: Should `ctx` always be immut? Rename this param/type?
+        _ = &mut ctx;
+        Ok(())
+    }
+
+    /// Performs factor-specific validation and configuration for the given
+    /// [`App`]. A runtime may - but is not required to - reuse the returned
+    /// config across multiple instances. Note that this may be called without
+    /// any call to `init` in cases where only validation is needed.
+    fn configure_app<T: RuntimeFactors>(
+        &self,
+        ctx: ConfigureAppContext<T>,
+        _runtime_config: &mut impl RuntimeConfig,
+    ) -> anyhow::Result<Self::AppState> {
+        _ = ctx;
+        Ok(Default::default())
```

Comment:
Just noting that I personally think it's worth the boiler plate of having implementations that don't require app configuration or instance preparation writing out these empty methods. 

It's a more verbose but arguably more clear (the reader instantly knows that they don't do any meaningful app configuration/instance preparation without having to know about the default implementation). And it prevents `AppState` and `InstancePreparer` types to always implement `Default` even when they don't have meaningful defaults.

Reply:
I've been trying to model how future changes to the `Factor` trait would work, but I suppose that doesn't mean we have to have default impls for _these_.

Reply:
I'm not sure I'm convinced about making `configure_app` mandatory yet but I've made the change and we can see how annoying it is as we implement more factors.

Reply:
I'm not sure I'm convinced about making `configure_app` mandatory yet but I've made the change and we can see how annoying it is as we implement more factors.